gprofiler/spark/spark_collector.py

```diff
@@ -0,0 +1,894 @@
+                running_apps[app_id] = (app_name, tracking_url)
+                # else:
+                #     # Only return the frameworks running on the correct port
+                #     tracking_url_port = urlparse(tracking_url).port
+                #     if tracking_url_port in spark_ports:
+                #         running_apps[app_id] = (app_name, tracking_url)
+        return running_apps
+
+    def _rest_request(self, url: str, object_path: str, *args: Any, **kwargs: Any) -> requests.Response:
+        """
+        Query the given URL and return the response
+        """
+        if object_path:
+            url = self._join_url_dir(url, object_path)
+
+        # Add args to the url
+        if args:
+            for directory in args:
+                url = self._join_url_dir(url, directory)
+
+        # Add kwargs as arguments
+        if kwargs:
+            query = "&".join(["{0}={1}".format(key, value) for key, value in kwargs.items() if value is not None])
+            url = urljoin(url, "?" + query)
+
+        try:
+            self._logger.debug(f"Spark check URL: {url}")
+            response = requests.get(url, timeout=3)
+            response.raise_for_status()
+
+            return response
+
+        except (HTTPError, InvalidURL, ConnectionError, ConnectionRefusedError, Timeout, ValueError):
+            raise
+
+    def _rest_request_to_json(self, address: str, object_path: str, *args: Any, **kwargs: Any) -> Any:
+        """
+        Query the given URL and return the JSON response
+        """
+
+        response = self._rest_request(address, object_path, *args, **kwargs)
+
+        try:
+            response_json = response.json()
+        except json.decoder.JSONDecodeError:
+            self._logger.exception("JSON Parse failed.")
+            raise
+
+        return response_json
+
+    def _get_request_url(self, url: str) -> str:
+        """
+        Get the request address, build with proxy if necessary
+        """
+        parsed = urlparse(url)
+
+        _url = url
+        if not (parsed.netloc and parsed.scheme):
+            _url = urljoin(self._master_address, parsed.path)
+
+        return _url
+
+    @staticmethod
+    def _join_url_dir(url: str, *args: Any) -> str:
+        """
+        Join a URL with multiple directories
+        """
+        for path in args:
+            url = url.rstrip("/") + "/"
+            url = urljoin(url, path.lstrip("/"))
+
+        return url
+
+
+class SparkSampler(object):
+    """
+    Spark cluster metrics sampler
+    """
+
+    def __init__(
+        self,
+        sample_period: float,
+        storage_dir: str,
+        api_client: Optional[APIClient] = None,
+    ):
+        self._master_address: Optional[str] = None
+        self._spark_mode: Optional[str] = None
+        self._collection_thread: Optional[Thread] = None
+        self._logger = get_logger_adapter(__name__)
+        self._sample_period = sample_period
+        self._stop_event = Event()
```

Comment:
Is it related to the profiler's `stop_event` used in other profilers?

If not - please add an explaining comment (to make it certain that this is meant to be a different object). How is the Spark Sampler stopped when I e.g use ^C to stop the agent?

If it is - use the same object as the rest of the profilers?

Reply:
Unrelated. What should the comment say? It is stopped when the `stop()` method is called by `GProfiler.stop()`.

Reply:
It just confused me that the entire project uses a `XXX._stop_event` object that refers to the same instance, and this class in particular uses another instance. I think it's nicer if they are the same instance, for consistency.

The comment could say "intentionally not the same instance as GProfiler._stop_event"

Reply:
My point is, a comment that won't leave a place for doubt "is it a mistake". It would have helped me read the code faster. Not critical